modules/javafx.controls/src/main/java/javafx/scene/control/DatePicker.java

```diff
@@ -121,8 +121,8 @@ public DatePicker(LocalDate localDate) {
             if (validateDate(chrono, date)) {
                 lastValidDate = date;
             } else {
-                System.err.println("Restoring value to " +
-                        ((lastValidDate == null) ? "null" : getConverter().toString(lastValidDate)));
+                //System.err.println("Restoring value to " +
+                //        ((lastValidDate == null) ? "null" : getConverter().toString(lastValidDate)));
                 setValue(lastValidDate);
```

Comment:
FYI I found it very weird that `DatePicker` tries to 'recover' itself when wrong values are set. I think this is the only? `Control` that does this behaviour

Reply:
I think ComboBox:243 also does that.

Basically, if we have a free-form text field as an input to a constrained value property, there must be a way to enforce the constraint.

Note that all this is not needed in e.g. ColorPicker because its "editor" always produces a valid value.